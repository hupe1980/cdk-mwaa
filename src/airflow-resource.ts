import * as path from 'node:path';
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as customResources from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';
import { Environment } from './environment';


export interface AirflowResourceBaseProps {
  readonly environment: Environment;
  readonly resourceType: string;
  readonly properties: Record<string, any>;
  readonly airflowRole?: string;
}


export abstract class AirflowResourceBase extends Construct {
  constructor(scope: Construct, id: string, props: AirflowResourceBaseProps) {
    super(scope, id);

    const { region, account } = cdk.Stack.of(this);
    const mwaaEnvName = props.environment.name;
    const airlfowRole = props.airflowRole ?? 'Op';

    // IAM Role for Lambda
    const lambdaRole = new iam.Role(this, 'Role', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
    });

    lambdaRole.addManagedPolicy(
      iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
    );

    lambdaRole.addToPolicy(
      new iam.PolicyStatement({
        actions: ['airflow:InvokeRestApi'],
        resources: [`arn:aws:airflow:${region}:${account}:role/${mwaaEnvName}/${airlfowRole}`],
      }),
    );

    // Lambda Function
    const airflowLambda = new lambda.SingletonFunction(this, 'Lambda', {
      uuid: `github.com/hupe1980/cdk-mwaa/airflow/${props.resourceType}`,
      runtime: lambda.Runtime.PYTHON_3_11,
      handler: `${props.resourceType}.on_event`,
      code: lambda.Code.fromAsset(path.join(__dirname, 'lambdas')),
      timeout: cdk.Duration.seconds(30),
      role: lambdaRole,
    });

    // Custom Resource Provider
    const provider = new customResources.Provider(this, 'Provider', {
      onEventHandler: airflowLambda,
      logGroup: new logs.LogGroup(this, 'ProviderLogGroup', {
        retention: logs.RetentionDays.ONE_DAY,
      }),
    });

    // Custom Resource to trigger Lambda
    new cdk.CustomResource(this, 'Resource', {
      serviceToken: provider.serviceToken,
      properties: {
        MwaaEnvName: mwaaEnvName,
        ...props.properties,
      },
    });

    // Add dependency on the MWAA environment
    this.node.addDependency(props.environment);
  }
}
