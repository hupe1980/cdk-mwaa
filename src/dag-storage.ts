import * as cdk from 'aws-cdk-lib';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as s3deploy from 'aws-cdk-lib/aws-s3-deployment';
import { Construct } from 'constructs';

/**
 * Configuration options for S3 path storage.
 */
export interface DagStorageConfigOptions {
  /** The S3 path where the resource is stored */
  readonly s3Path: string;
  /** Optional object version for versioned buckets */
  readonly s3ObjectVersion?: string;
}

/**
 * Options for deploying files to the DAG storage bucket.
 */
export interface DagStorageDeployOptions {
  /** Sources to deploy into the bucket */
  readonly sources?: s3deploy.ISource[];
  /** Whether to retain files on deletion of the stack */
  readonly retainOnDelete?: boolean;
  /** Whether to prune old versions of deployed files */
  readonly prune?: boolean;
}

/**
 * Properties for configuring the DAG storage bucket.
 */
export interface DagStorageProps {
  /** Optional custom bucket name */
  readonly bucketName?: string;
  /** Enable versioning for the bucket */
  readonly versioned?: boolean;
  /** Lifecycle rule for expiring non-current versions */
  readonly noncurrentVersionExpiration?: cdk.Duration;
  /** Policy for bucket removal */
  readonly removalPolicy?: cdk.RemovalPolicy;
  /** Path for storing DAG files */
  readonly dagS3Path?: string;
  /** Configuration for plugins storage */
  readonly pluginsConfig?: DagStorageConfigOptions;
  /** Configuration for requirements storage */
  readonly requirementsConfig?: DagStorageConfigOptions;
  /** Configuration for startup script storage */
  readonly startupScriptConfig?: DagStorageConfigOptions;
  /** Options for deploying files into the bucket */
  readonly deployOptions?: DagStorageDeployOptions;
}

/**
 * Represents a storage location for MWAA DAGs and dependencies in an S3 bucket.
 */
export class DagStorage extends Construct {
  /** The S3 bucket storing DAGs, plugins, requirements, and startup scripts */
  public readonly bucket: s3.Bucket;
  /** S3 path for DAGs */
  public readonly dagS3Path?: string;
  /** Plugin storage configuration */
  public readonly pluginsConfig?: DagStorageConfigOptions;
  /** Requirements storage configuration */
  public readonly requirementsConfig?: DagStorageConfigOptions;
  /** Startup script storage configuration */
  public readonly startupScriptConfig?: DagStorageConfigOptions;

  constructor(scope: Construct, id: string, props: DagStorageProps) {
    super(scope, id);

    // Create S3 bucket with defined properties
    this.bucket = new s3.Bucket(this, 'DagStorageBucket', {
      bucketName: props.bucketName,
      versioned: props.versioned,
      encryption: s3.BucketEncryption.S3_MANAGED,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      removalPolicy: props.removalPolicy ?? cdk.RemovalPolicy.RETAIN,
      autoDeleteObjects: props.removalPolicy === cdk.RemovalPolicy.DESTROY,
    });

    // Apply lifecycle rule if versioning and expiration settings are provided
    if (props.versioned && props.noncurrentVersionExpiration) {
      this.bucket.addLifecycleRule({
        id: 'DeleteOldVersions',
        noncurrentVersionExpiration: props.noncurrentVersionExpiration,
      });
    }

    // Assign configuration properties
    this.dagS3Path = props.dagS3Path ?? 'dags/';
    this.pluginsConfig = props.pluginsConfig;
    this.requirementsConfig = props.requirementsConfig;
    this.startupScriptConfig = props.startupScriptConfig;

    // Deploy specified sources to the bucket if provided
    if (props.deployOptions?.sources && props.deployOptions?.sources.length > 0) {
      new s3deploy.BucketDeployment(this, 'DagStorageDeployment', {
        sources: props.deployOptions.sources,
        destinationBucket: this.bucket,
        retainOnDelete: props.deployOptions.retainOnDelete ?? false,
        prune: props.deployOptions.prune ?? true,
      });
    }
  }
}
