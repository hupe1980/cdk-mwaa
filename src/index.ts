export * from './dag-storage';
export * from './sizing';
export * from './environment';
export * from './patterns';
export * from './security-group';