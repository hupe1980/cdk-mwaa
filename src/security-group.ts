import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

/**
 * Properties for defining a Security Group.
 */
export interface SecurityGroupProps extends ec2.SecurityGroupProps {}

/**
 * A custom Security Group with a self-referencing ingress rule for MWAA.
 */
export class SecurityGroup extends ec2.SecurityGroup {
  /**
   * Creates a new Security Group with self-referencing ingress rules.
   *
   * @param scope The parent construct.
   * @param id The unique identifier for this construct.
   * @param props Security group properties.
   */
  constructor(scope: Construct, id: string, props: SecurityGroupProps) {
    super(scope, id, props);

    // Allow all inbound traffic within the security group
    this.addIngressRule(this, ec2.Port.allTraffic(), 'MWAA Self-referencing rule');
  }
}
