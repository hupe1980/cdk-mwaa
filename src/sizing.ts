/**
 * Represents the available environment classes for MWAA (Managed Workflows for Apache Airflow).
 */
export enum EnvironmentClass {
  MW1_MICRO = 'mw1.micro',
  MW1_SMALL = 'mw1.small',
  MW1_MEDIUM = 'mw1.medium',
  MW1_LARGE = 'mw1.large',
}

/**
 * Defines the configuration properties for sizing an MWAA environment.
 */
export interface SizingProps {
  /** The environment class determining the available resources. */
  readonly environmentClass: EnvironmentClass;

  /** Minimum number of workers in the MWAA environment. */
  readonly minWorkers: number;

  /** Maximum number of workers in the MWAA environment. */
  readonly maxWorkers: number;

  /** Minimum number of webservers in the MWAA environment. */
  readonly minWebservers: number;

  /** Maximum number of webservers in the MWAA environment. */
  readonly maxWebservers: number;

  /** Number of schedulers in the MWAA environment. */
  readonly schedulers: number;
}

/**
 * Provides predefined and customizable sizing options for an MWAA environment.
 */
export class Sizing {
  /**
   * Creates an MW1_MICRO sized environment with a single worker, webserver, and scheduler.
   */
  static mw1Micro(): Sizing {
    return new Sizing({
      environmentClass: EnvironmentClass.MW1_MICRO,
      minWorkers: 1,
      maxWorkers: 1,
      minWebservers: 1,
      maxWebservers: 1,
      schedulers: 1,
    });
  }

  /**
   * Creates an MW1_SMALL sized environment with a predefined range of workers and webservers.
   */
  static mw1Small(): Sizing {
    return new Sizing({
      environmentClass: EnvironmentClass.MW1_SMALL,
      minWorkers: 2,
      maxWorkers: 5,
      minWebservers: 2,
      maxWebservers: 5,
      schedulers: 2,
    });
  }

  /**
   * Creates an MW1_MEDIUM sized environment with a predefined range of workers and webservers.
   */
  static mw1Medium(): Sizing {
    return new Sizing({
      environmentClass: EnvironmentClass.MW1_MEDIUM,
      minWorkers: 2,
      maxWorkers: 5,
      minWebservers: 2,
      maxWebservers: 5,
      schedulers: 2,
    });
  }

  /**
   * Creates an MW1_LARGE sized environment with a predefined range of workers and webservers.
   */
  static mw1Large(): Sizing {
    return new Sizing({
      environmentClass: EnvironmentClass.MW1_LARGE,
      minWorkers: 2,
      maxWorkers: 5,
      minWebservers: 2,
      maxWebservers: 5,
      schedulers: 2,
    });
  }

  /**
   * Creates a custom-sized MWAA environment based on user-defined configuration.
   *
   * @param config Custom sizing properties.
   */
  static custom(config: SizingProps): Sizing {
    return new Sizing(config);
  }

  /**
   * Creates a new instance of Sizing.
   *
   * @param props The sizing properties for the MWAA environment.
   */
  private constructor(
    private readonly props: SizingProps,
  ) {
    this.validateConfig();
  }

  /**
   * Validates the sizing configuration to ensure all values are within allowed ranges.
   */
  private validateConfig(): void {
    const { environmentClass, minWorkers, maxWorkers, minWebservers, maxWebservers, schedulers } = this.props;

    if (environmentClass === EnvironmentClass.MW1_MICRO) {
      if (minWorkers !== 1 || maxWorkers !== 1 || minWebservers !== 1 || maxWebservers !== 1 || schedulers !== 1) {
        throw new Error('For MW1_MICRO environment class, minWorkers, maxWorkers, minWebservers, maxWebservers, and schedulers must all be 1.');
      }
    } else {
      if (minWorkers < 1 || minWorkers > 25) {
        throw new Error('For non-MW1_MICRO environment classes, minWorkers must be between 1 and 25.');
      }
      if (maxWorkers < 1 || maxWorkers > 25) {
        throw new Error('For non-MW1_MICRO environment classes, maxWorkers must be between 1 and 25.');
      }
      if (minWorkers > maxWorkers) {
        throw new Error('For non-MW1_MICRO environment classes, minWorkers must be less than or equal to maxWorkers.');
      }
      if (minWebservers < 2 || minWebservers > 5) {
        throw new Error('For non-MW1_MICRO environment classes, minWebservers must be between 2 and 5.');
      }
      if (maxWebservers < 2 || maxWebservers > 5) {
        throw new Error('For non-MW1_MICRO environment classes, maxWebservers must be between 2 and 5.');
      }
      if (minWebservers > maxWebservers) {
        throw new Error('For non-MW1_MICRO environment classes, minWebservers must be less than or equal to maxWebservers.');
      }
      if (schedulers < 2 || schedulers > 5) {
        throw new Error('For non-MW1_MICRO environment classes, schedulers must be between 2 and 5.');
      }
    }
  }

  /** Returns the environment class. */
  get environmentClass(): EnvironmentClass {
    return this.props.environmentClass;
  }

  /** Returns the minimum number of workers. */
  get minWorkers(): number {
    return this.props.minWorkers;
  }

  /** Returns the maximum number of workers. */
  get maxWorkers(): number {
    return this.props.maxWorkers;
  }

  /** Returns the minimum number of webservers. */
  get minWebservers(): number {
    return this.props.minWebservers;
  }

  /** Returns the maximum number of webservers. */
  get maxWebservers(): number {
    return this.props.maxWebservers;
  }

  /** Returns the number of schedulers. */
  get schedulers(): number {
    return this.props.schedulers;
  }
}
