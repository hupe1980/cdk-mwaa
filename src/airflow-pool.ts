import { Construct } from 'constructs';
import { AirflowResourceBase } from './airflow-resource';
import { Environment } from './environment';

export interface AirflowPoolProps {
  readonly environment: Environment;
  readonly poolName: string;
  readonly poolSlots?: number;
  readonly poolDescription?: string;
}

export class AirflowPool extends AirflowResourceBase {
  constructor(scope: Construct, id: string, props: AirflowPoolProps) {
    super(scope, id, {
      environment: props.environment,
      resourceType: 'pool',
      properties: {
        PoolName: props.poolName,
        PoolSlots: props.poolSlots ?? 3,
        PoolDescription: props.poolDescription ?? '',
      },
    });
  }
}
