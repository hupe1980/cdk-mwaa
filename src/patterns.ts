import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

import { DagStorage, DagsOptions, ConfigsOptions } from './dag-storage';
import { Environment, WebserverAccessMode } from './environment';
import { Sizing } from './sizing';

/**
 * Interface defining the properties for configuring MWAA (Managed Airflow).
 */
export interface MWAAProps {
  /**
   * The name of the Airflow environment.
   */
  readonly environmentName: string;

  /**
   * The version of Airflow to deploy.
   */
  readonly airflowVersion: string;

  /**
   * Optional sizing configuration for the MWAA environment.
   * Defines the compute resources.
   */
  readonly sizing?: Sizing;

  /**
   * The VPC in which to deploy the MWAA environment.
   * If not provided, a default VPC will be created.
   */
  readonly vpc?: ec2.IVpc;

  /**
   * The name of the S3 bucket used for storing DAGs.
   * If not provided, a default bucket is created.
   */
  readonly bucketName?: string;

  /** Configuration for DAG storage. */
  readonly dagsOptions?: DagsOptions;

  /** Configuration for plugins storage. */
  readonly configsOptions?: ConfigsOptions;

  /**
   * Airflow configuration options as key-value pairs.
   * These configuration options are passed to the Airflow environment.
   */
  readonly airflowConfigurationOptions?: { [key: string]: any };

  /**
   * The removal policy for the MWAA resources.
   * Determines what happens to the resources when they are deleted.
   * Defaults to 'RETAIN' if not specified.
   */
  readonly removalPolicy?: cdk.RemovalPolicy;
}

/**
 * PublicRoutingMWAA constructs a Managed Airflow (MWAA) environment with public webserver access.
 * It creates the necessary VPC, S3 storage for DAGs, and an Airflow environment.
 */
export class PublicRoutingMWAA extends Construct {
  constructor(scope: Construct, id: string, props: MWAAProps) {
    super(scope, id);

    // Create a VPC if not provided, with specific configuration for subnets and NAT Gateways
    const vpc = props.vpc ?? new ec2.Vpc(this, 'Vpc', {
      maxAzs: 2, // Use up to 2 Availability Zones
      natGateways: 2, // Use 2 NAT Gateways for internet access
      createInternetGateway: true, // Create an Internet Gateway for public access
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'public-subnet',
          subnetType: ec2.SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'private-subnet',
          subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
        },
      ],
    });

    // Create the DAG storage (S3 bucket) for the MWAA environment
    const dagStorage = new DagStorage(this, 'DagStorage', {
      bucketName: props.bucketName,
      dagsOptions: props.dagsOptions,
      configsOptions: props.configsOptions,
      removalPolicy: props.removalPolicy ?? cdk.RemovalPolicy.RETAIN,
    });

    // Create the Airflow environment
    const environment = new Environment(this, 'Environment', {
      name: props.environmentName,
      dagStorage,
      vpc,
      webserverAccessMode: WebserverAccessMode.PUBLIC_ONLY, // Public access to the webserver
      airflowVersion: props.airflowVersion,
      sizing: props.sizing ?? Sizing.mw1Micro(), // Default to 'mw1Micro' sizing if not provided
      airflowConfigurationOptions: props.airflowConfigurationOptions,
    });

    // Enable the secrets backend for Airflow (e.g., for environment variables or secrets storage)
    environment.enableSecretsBackend();
  }
}

// TODO: Refactor this class to implement private routing, if required
// export class PrivateRoutingMWAA extends Construct {
//   constructor(scope: Construct, id: string, props: MWAAProps) {
//     super(scope, id);
//     // Implement private routing logic for MWAA here
//   }
// }
