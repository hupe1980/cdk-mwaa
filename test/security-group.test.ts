import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { SecurityGroup } from '../src';

describe('SecurityGroup Construct', () => {
  let stack: cdk.Stack;
  let vpc: ec2.Vpc;

  beforeEach(() => {
    stack = new cdk.Stack();
    vpc = new ec2.Vpc(stack, 'TestVpc');
  });

  test('SecurityGroup should be created with correct properties', () => {
    new SecurityGroup(stack, 'TestSecurityGroup', {
      vpc,
      description: 'Test security group',
      allowAllOutbound: true,
    });

    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
  });

  test('SecurityGroup should have a self-referencing ingress rule', () => {
    new SecurityGroup(stack, 'TestSecurityGroup', {
      vpc,
    });

    const template = Template.fromStack(stack);
    template.hasResourceProperties('AWS::EC2::SecurityGroupIngress', {
      IpProtocol: '-1',
      SourceSecurityGroupId: {
        'Fn::GetAtt': ['TestSecurityGroup880B57C0', 'GroupId'],
      },
    });
  });
});
