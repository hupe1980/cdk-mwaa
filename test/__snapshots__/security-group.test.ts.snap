// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecurityGroup Construct SecurityGroup should be created with correct properties 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TestSecurityGroup880B57C0": {
      "Properties": {
        "GroupDescription": "Test security group",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "TestSecurityGroupfromTestSecurityGroupALLTRAFFIC2EA53DBD": {
      "Properties": {
        "Description": "MWAA Self-referencing rule",
        "GroupId": {
          "Fn::GetAtt": [
            "TestSecurityGroup880B57C0",
            "GroupId",
          ],
        },
        "IpProtocol": "-1",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "TestSecurityGroup880B57C0",
            "GroupId",
          ],
        },
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "TestVpcE77CE678": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/TestVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "TestVpcIGW9DD53F70": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/TestVpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "TestVpcPrivateSubnet1DefaultRoute32E7B814": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "TestVpcPublicSubnet1NATGatewayA323E3EC",
        },
        "RouteTableId": {
          "Ref": "TestVpcPrivateSubnet1RouteTable469B0105",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVpcPrivateSubnet1RouteTable469B0105": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/TestVpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVpcPrivateSubnet1RouteTableAssociationFFD4DFF7": {
      "Properties": {
        "RouteTableId": {
          "Ref": "TestVpcPrivateSubnet1RouteTable469B0105",
        },
        "SubnetId": {
          "Ref": "TestVpcPrivateSubnet1SubnetCC65D771",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVpcPrivateSubnet1SubnetCC65D771": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "Default/TestVpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVpcPrivateSubnet2DefaultRouteA7EB6930": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "TestVpcPublicSubnet2NATGatewayA9858C31",
        },
        "RouteTableId": {
          "Ref": "TestVpcPrivateSubnet2RouteTableCEF29F7C",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVpcPrivateSubnet2RouteTableAssociation18250AB4": {
      "Properties": {
        "RouteTableId": {
          "Ref": "TestVpcPrivateSubnet2RouteTableCEF29F7C",
        },
        "SubnetId": {
          "Ref": "TestVpcPrivateSubnet2SubnetDE0C64A2",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVpcPrivateSubnet2RouteTableCEF29F7C": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/TestVpc/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVpcPrivateSubnet2SubnetDE0C64A2": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.192.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "Default/TestVpc/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVpcPublicSubnet1DefaultRoute6C0F0315": {
      "DependsOn": [
        "TestVpcVPCGWF1827B84",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "TestVpcIGW9DD53F70",
        },
        "RouteTableId": {
          "Ref": "TestVpcPublicSubnet1RouteTable4CBFF871",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVpcPublicSubnet1EIP4884338C": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/TestVpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "TestVpcPublicSubnet1NATGatewayA323E3EC": {
      "DependsOn": [
        "TestVpcPublicSubnet1DefaultRoute6C0F0315",
        "TestVpcPublicSubnet1RouteTableAssociation7D1DECD9",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "TestVpcPublicSubnet1EIP4884338C",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "TestVpcPublicSubnet1SubnetA7DB1EDF",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/TestVpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "TestVpcPublicSubnet1RouteTable4CBFF871": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/TestVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVpcPublicSubnet1RouteTableAssociation7D1DECD9": {
      "Properties": {
        "RouteTableId": {
          "Ref": "TestVpcPublicSubnet1RouteTable4CBFF871",
        },
        "SubnetId": {
          "Ref": "TestVpcPublicSubnet1SubnetA7DB1EDF",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVpcPublicSubnet1SubnetA7DB1EDF": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "Default/TestVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVpcPublicSubnet2DefaultRoute054DAE0A": {
      "DependsOn": [
        "TestVpcVPCGWF1827B84",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "TestVpcIGW9DD53F70",
        },
        "RouteTableId": {
          "Ref": "TestVpcPublicSubnet2RouteTable75B88314",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVpcPublicSubnet2EIP83F7944C": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/TestVpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "TestVpcPublicSubnet2NATGatewayA9858C31": {
      "DependsOn": [
        "TestVpcPublicSubnet2DefaultRoute054DAE0A",
        "TestVpcPublicSubnet2RouteTableAssociationB386A819",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "TestVpcPublicSubnet2EIP83F7944C",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "TestVpcPublicSubnet2Subnet80A14523",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/TestVpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "TestVpcPublicSubnet2RouteTable75B88314": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/TestVpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVpcPublicSubnet2RouteTableAssociationB386A819": {
      "Properties": {
        "RouteTableId": {
          "Ref": "TestVpcPublicSubnet2RouteTable75B88314",
        },
        "SubnetId": {
          "Ref": "TestVpcPublicSubnet2Subnet80A14523",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVpcPublicSubnet2Subnet80A14523": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "Default/TestVpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVpcVPCGWF1827B84": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "TestVpcIGW9DD53F70",
        },
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
